### PR TITLE
added LATEST and EARLIEST group functions for DateTime groups

### PR DESCRIPTION
**Pull request by rmichalak**
Fixes: #5177
Signed-off-by: Robert Michalak <rbrt.michalak@gmail.com>

### CODE REVIEW

**Comment by htreu** [comment]
Hi @rmichalak thanks a lot for this proposal.

Taken from #5177:

> MAX/MIN group functions are limited to decimal types and extending them for that purpose does not seem like a good idea.

Why not extending MIN/MAX? The mapping is obvious and we would have the same functionality with less code.
For Paper UI there is a base item specific selection of group functions when creating a new group. Please see [controllers.items.js#262](https://github.com/eclipse/smarthome/blob/master/extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/items/controllers.items.js#L262) for details and extend it for DateTime, too. 

Also: The new class `DateTimeGroupFunction` is missing in the commit.

**Comment by rmichalak** [comment]
Hi @htreu,

Whoops! missing file... I'm ashamed... Will fill that gap in the evening.
My initial idea was to extend MIN/MAX for that purpose, but what about groups containing both Numeric and DateTime items? We could in theory base our logic on group type and perform calculation either on DateTime or Numeric items, but AFAIK we don't have access to this information inside `calculate` method(?).
This is why I decided to create new functions, dedicated to DateTime types.

**Comment by htreu** [comment]
> We could in theory base our logic on group type and perform calculation either on DateTime or Numeric items, but AFAIK we don't have access to this information inside calculate method(?).

Thats indeed a good argument! We could apply some heuristic what MIN/MAX would probably mean but thats hard to understand/explain and may in the end be wrong anyhow.
Please push the missing class and have a look at the Paper UI group function selection. There is another file [shared.properties.js](https://github.com/eclipse/smarthome/blob/master/extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/shared.properties.js#L148) which you might need to extend for the DateTime function selection.

**Comment by rmichalak** [comment]
Sure, will do. I will also take a closer look at PaperUI, cause this part of the project was off my radar so far.
Thanks for your suggestions.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeGroupFunction.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import java.time.ZonedDateTime;
+import java.util.Set;
+
+import org.eclipse.smarthome.core.items.GroupFunction;
+import org.eclipse.smarthome.core.items.Item;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+
+/**
+ * This interface is a container for dimension based functions that require {@link DateTypeType}s for its calculations.
```
Comment (htreu) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeGroupFunction.java]: `for dimension based functions`
These group functions are not dimension based but specifically designed for the DateTimeItem base item type.

Comment (rmichalak) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeGroupFunction.java]: Good catch. Too many copy-pastes ;). Fixed.

**Comment by rmichalak** [comment]
It looks like CI failure is not related to my changes. Is there a way to run it again?

**Comment by kaikreuzer** [comment]
I retriggered Travis.

**Pull request merged**